File: backend/services/bootmode/bootmode.go
// Package bootmode reports THIS INSTANCE's own identity-and-replication state,
// and serves it at GET /api/setup/mode.
//
// # What this package cannot tell you
//
// It cannot tell you whether a human has completed first-run setup on this box.
// GET /api/setup/status is the sole authority for that: it stats the
// setup-complete marker written by POST /api/setup/complete
// (backend/cmd/server/routes_setup.go). Nothing here reads that marker, and
// nothing here should be used as a stand-in for it.
//
// That sentence is here because the opposite assumption shipped. The modes used
// to be named "setup", "sync" and "normal", and the wizard read "normal" as
// "already set up":
//
//	if (data.mode === 'normal') {
//	  // Already set up — shouldn't be here, but complete gracefully
//	  onComplete()
//	}
//
// "normal" only ever meant "db/instance.json exists and no sync is running".
// registerIdentityRoutes calls identity.Load at STARTUP, and identity.Load
// creates instance.json when it is missing — so the file exists before the
// server accepts its first connection. The old "normal" was therefore true on
// every first boot, before any human had done anything, and the wizard
// dismissed itself into the login screen on a box with no accounts. Worse, the
// old "setup" mode was unreachable over HTTP for the same reason: no client
// could ever observe it. TestModeInstanceAbsentIsUnreachableOverHTTP
// (backend/cmd/server/routes_bootmode_reachability_test.go) pins that.
//
// # The modes
//
// The names now say what is on disk, so there is no word left to misread as a
// statement about the owner:
//
//	instance_absent — db dir or db/instance.json is missing: this process has
//	                  not yet written its own instance identity. Observable only
//	                  in-process (a server that is serving has already written
//	                  it), and NOT a synonym for "needs setup".
//	syncing         — db/sync-state.json says status "syncing": this box is
//	                  replicating from a cluster it is joining.
//	instance_ready  — instance identity present, no replication in progress.
//	                  Says NOTHING about accounts, ownership, or setup.
//
// The only question a caller can honestly answer from this endpoint is "is this
// box mid-join?" — i.e. mode == syncing. frontend/src/lib/bootmode.ts mirrors
// these three strings and is checked against this file by TestModeStringsMatchFrontend.
package bootmode

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
)

// The three mode strings this package can emit. Callers must branch on these
// constants rather than on string literals, and must not infer setup state from
// any of them — see the package doc.
const (
	// ModeInstanceAbsent: no db dir, or no db/instance.json. The process has
	// not written its instance identity yet.
	ModeInstanceAbsent = "instance_absent"
	// ModeSyncing: db/sync-state.json reports status "syncing".
	ModeSyncing = "syncing"
	// ModeInstanceReady: instance identity on disk, no replication running.
	ModeInstanceReady = "instance_ready"
)

// Modes returns every value Detect can put in Result.Mode, in the order they
// are documented. Tests use it to enumerate the wire contract; nothing should
// hard-code a subset of it.
func Modes() []string {
	return []string{ModeInstanceAbsent, ModeSyncing, ModeInstanceReady}
}

// Result holds the detected instance state and optional sync state.
type Result struct {
	Mode      string `json:"mode"`
	SyncState string `json:"sync_state,omitempty"`
}

// syncStateFile is the filename within the db dir that tracks cluster sync progress.
const syncStateFile = "sync-state.json"

// instanceFile is the filename that records a persisted instance identity (NET-06).
const instanceFile = "instance.json"

// Detect inspects home to determine this instance's state.
// home is the box data root; the db dir is home/db.
func Detect(home string) (Result, error) {
	dbDir := filepath.Join(home, "db")

	// Rule 1: db dir absent → no instance identity yet.
	if _, err := os.Stat(dbDir); os.IsNotExist(err) {
		return Result{Mode: ModeInstanceAbsent}, nil
	} else if err != nil {
		return Result{}, err
	}

	// Rule 2: instance.json absent → no instance identity yet.
	instancePath := filepath.Join(dbDir, instanceFile)
	if _, err := os.Stat(instancePath); os.IsNotExist(err) {
		return Result{Mode: ModeInstanceAbsent}, nil
	} else if err != nil {
		return Result{}, err
	}

	// Rule 3: sync-state.json present with status "syncing" → replicating.
	syncPath := filepath.Join(dbDir, syncStateFile)
	if data, err := os.ReadFile(syncPath); err == nil {
		var ss struct {
			Status string `json:"status"`
		}
		if json.Unmarshal(data, &ss) == nil && ss.Status == "syncing" {
			return Result{Mode: ModeSyncing, SyncState: ss.Status}, nil
		}
	}

	// Rule 4: identity present, nothing replicating.
	return Result{Mode: ModeInstanceReady}, nil
}

// HasInstanceIdentity reports whether this box has written its own instance
// identity to disk.
//
// This is a statement about ONE FILE, db/instance.json. It is true on a
// pristine, unconfigured, account-less box the moment the server starts, so it
// must never be used to gate anything on "the owner has set this box up" —
// that question belongs to the setup-complete marker.
func HasInstanceIdentity(home string) bool {
	result, err := Detect(home)
	if err != nil {
		return false
	}
	return result.Mode == ModeInstanceReady || result.Mode == ModeSyncing
}

// RegisterHandlers registers the instance-state HTTP endpoint on mux.
// GET /api/setup/mode returns JSON
// {"mode":"instance_absent"|"syncing"|"instance_ready","sync_state":"..."}.
func RegisterHandlers(mux *http.ServeMux, home string) {
	mux.HandleFunc("GET /api/setup/mode", func(w http.ResponseWriter, r *http.Request) {
		result, err := Detect(home)
		if err != nil {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusInternalServerError)
			json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
			return
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(result)
	})
}

File: frontend/src/lib/bootmode.ts
/**
 * bootmode.ts — the shell's mirror of GET /api/setup/mode.
 *
 * This endpoint reports the BOX'S OWN instance-identity and replication state.
 * It does not answer "has the owner finished first-run setup?" — GET
 * /api/setup/status does, and AuthGate has already asked it before anything
 * that imports this file is on screen (App.tsx / lib/setupProbe.ts).
 *
 * The distinction is not academic. The three modes used to be named
 * `setup` / `sync` / `normal`, and Setup.tsx read `normal` as "already set up"
 * and dismissed itself:
 *
 *     if (data.mode === 'normal') {
 *       // Already set up — shouldn't be here, but complete gracefully
 *       onComplete()
 *     }
 *
 * `normal` meant "db/instance.json exists and nothing is syncing". The server
 * writes instance.json at STARTUP (registerIdentityRoutes → identity.Load), so
 * it was already true on a box nobody had ever touched. The fifteen-step wizard
 * therefore mounted, asked one question of the box, and handed the founder a
 * "Create your account" login form on a box with no accounts — the exact
 * symptom a backend fix had just been shipped to kill.
 *
 * So: the only question this endpoint can honestly answer is the one below.
 * Everything else about it is state that no first-run decision may depend on.
 *
 * The strings are the Go package's, checked against
 * backend/services/bootmode/bootmode.go by TestModeStringsMatchFrontend — a
 * value added on one side and not the other fails the Go build's test run.
 */

/** No db dir or no db/instance.json yet. Unreachable over HTTP: the server has
 *  written instance.json before it accepts a connection. Listed because the
 *  endpoint's contract includes it, not because a browser can see it. */
export const MODE_INSTANCE_ABSENT = 'instance_absent'

/** db/sync-state.json says "syncing": this box is replicating from a cluster. */
export const MODE_SYNCING = 'syncing'

/** Instance identity on disk, nothing replicating. TRUE ON A PRISTINE FIRST
 *  BOOT. Says nothing about accounts, ownership, or setup. */
export const MODE_INSTANCE_READY = 'instance_ready'

/** Every value the endpoint can emit. */
export const INSTANCE_MODES = [
  MODE_INSTANCE_ABSENT,
  MODE_SYNCING,
  MODE_INSTANCE_READY,
] as const

export type InstanceMode = (typeof INSTANCE_MODES)[number]

function isRecord(x: unknown): x is Record<string, unknown> {
  return typeof x === 'object' && x !== null
}

/**
 * Is this box currently replicating from a cluster it is joining?
 *
 * The ONLY thing a caller may conclude from /api/setup/mode. Anything that is
 * not an explicit `syncing` — a different mode, a malformed body, a captive
 * portal's HTML, a 500, an unreachable box — is `false`, i.e. "carry on with
 * the normal flow". That default is deliberate and it is the safe one: the
 * normal flow is the fifteen-step wizard, which the caller only reached because
 * /api/setup/status said setup was outstanding.
 */
export function isJoiningCluster(data: unknown): boolean {
  return isRecord(data) && data.mode === MODE_SYNCING
}
